### PR TITLE
Source files for vendored libraries

### LINKED ISSUE #1126: Fix installation if libraries are missing
**What is the feature or improvement you would like to see?**

We're declaring libraries as optional, can we add detection and fallback code to `configure` to actually live by it?

https://github.com/igraph/rigraph/blob/bf64f6a7cd97e83c0622f37c4107d8c81194d6ec/DESCRIPTION#L73-L74

If all dependencies are optional, installation should work on any system regarding of system libraries. If not, we need to document it.

**Use cases for the feature**

Users keep reporting installation failures when installing from source. This is something I'd really like to see in the the upcoming release.

### PR DESCRIPTION
**Pull request by Antonov548**
Closes #1126.

### CODE REVIEW

**Comment by Antonov548** [comment]
> Thanks. I'll review, can we discuss at around 16:00 if needed?

Yes, sure. I trying to find a way to check that glpk in installed, but seems there is no straight forward solution.

File: configure [hunk 1]
```diff
@@ -12,15 +12,26 @@ else
...
 fi
 
+OBJECTS='${SOURCES}'
+
+if echo '#include <glpk.h>' | $(R CMD config CC) -E $CPPFLAGS - > /dev/null 2>&1; then
+  echo "Using installed GLPK"
+  PKG_LIBS="$PKG_LIBS -lglpk"
+else
+  echo "Using vendored GLPK"
+  PKG_CFLAGS="$PKG_CFLAGS -Ivendor/cigraph/vendor/glpk -Ivendor/cigraph/vendor/glpk/env -Ivendor/cigraph/vendor/glpk/minisat -Ivendor/cigraph/vendor/glpk/misc -Ivendor/cigraph/vendor/glpk/draft -Ivendor/cigraph/vendor/glpk/npp -Ivendor/cigraph/vendor/glpk/api -Ivendor/cigraph/vendor/glpk/mpl -Ivendor/cigraph/vendor/glpk/bflib -Ivendor/cigraph/vendor/glpk/amd -Ivendor/cigraph/vendor/glpk/simplex -Ivendor/cigraph/vendor/glpk/colamd"
```
Comment (szhorvat) [inline configure:29]: `-I` flags should go in CPPFLAGS instead of CFLAGS, no?

File: src/Makevars.in [hunk 2]
```diff
@@ -1,14 +1,16 @@
 include sources.mk
+include vendor/cigraph/vendor/glpk/sources-glpk.mk
 
 PKG_CFLAGS=$(C_VISIBILITY)
 PKG_CXXFLAGS=$(CXX_VISIBILITY)
 PKG_FFLAGS=$(F_VISIBILITY)
 
-PKG_CPPFLAGS=-DUSING_R -I. -Ivendor -Ivendor/cigraph/src -Ivendor/cigraph/include -Ivendor/cigraph/vendor @cflags@ \
+# FIXME: Make INTERNAL_GMP depending on availability of gmp.h
+PKG_CPPFLAGS=-DUSING_R -DINTERNAL_GMP=1 -I. -Ivendor -Ivendor/cigraph/src -Ivendor/cigraph/include -Ivendor/cigraph/vendor @cflags@ \
```
Comment (szhorvat) [inline src/Makevars.in:9]: `-DINTERNAL_GMP` is sufficient, no need for `=1` (but not is it wrong).

File: src/Makevars.in [hunk 3]
```diff
@@ -1,14 +1,16 @@
 include sources.mk
+include vendor/cigraph/vendor/glpk/sources-glpk.mk
 
 PKG_CFLAGS=$(C_VISIBILITY)
 PKG_CXXFLAGS=$(CXX_VISIBILITY)
 PKG_FFLAGS=$(F_VISIBILITY)
 
-PKG_CPPFLAGS=-DUSING_R -I. -Ivendor -Ivendor/cigraph/src -Ivendor/cigraph/include -Ivendor/cigraph/vendor @cflags@ \
+# FIXME: Make INTERNAL_GMP depending on availability of gmp.h
```
Comment (szhorvat) [inline src/Makevars.in:8]: Using an external GMP provides no real benefit at the moment, other than not having to deal with CRAN's complaints about compiler warnings.

Attention: Do not patch `mini-gmp.c` to eliminate warnings! GMP is too easy to break accidentally.

**Comment by Antonov548** [comment]
> @Antonov548 is any help needed to fix the conflicts?

I think it's better to start from scratch. Since part was already implemented in https://github.com/igraph/rigraph/pull/1233

**Comment by maelle** [comment]
Ok so you want to close this PR and start a new one? I'm a bit lost as I don't know how you're generating the PR. (but I can learn!)

**Comment by Antonov548** [comment]
> Ok so you want to close this PR and start a new one? I'm a bit lost as I don't know how you're generating the PR. (but I can learn!)

Yes, I will create new one. Just need to clarify what still needs to be done regarding vendored libraries, since as I mentioned part is already done.

File: configure (hunk 1 above)
Comment (Antonov548) [reply configure:38]: Is this comment just about naming? As I understand it doesn't has functional meaning.

**Comment by Antonov548** [comment]
@krlmlr Seems it couldn't detect `glpk` on pipeline just with commpilation single file. Not sure why it happens. But using vendored `glpk` also doesn't work properlly now. Not all files from `vendor/cigraph/vendor/glpk` should be used.

File: configure (hunk 1 above)
Comment (krlmlr) [reply configure:38]: We can leave it as is and resolve later if needed.

**Comment by Antonov548** [comment]
Oh, I missed. It's has errors in checking code rather than build the package. Can we try to avoid this issues with not including files which raise the warnings?

**Comment by Antonov548** [comment]
@krlmlr is there any way to ingore warnings in vendored folder for the check step?

**Comment by krlmlr** [comment]
@Antonov548: Turning off warnings for certain parts of the code will be difficult.

@szhorvat: How did this even work with pre-2.0.0 without GLPK installed? Did `R CMD check` complain then?

Only CRAN is wary about these warnings, but they will never see them because they have GLPK installed. (Or so I hope.) We could just ignore those warnings on GHA, but we need multiple runs then: one with, and one without GLPK.

**Comment by szhorvat** [comment]
> How did this even work with pre-2.0.0 without GLPK installed?

It disabled GLPK.

Do we need to care about R CMD check for anything except CRAN? CRAN has GLPK.

**Comment by szhorvat** [comment]
I should point out that we go to great lengths to make sure that igraph doesn't print without permission. So the complaints about `printf` being referenced are theoretical, and only relevant for avoiding an argument with CRAN.

As for patching out all the `printf`s in the vendored GLPK: we don't do that because it makes it much more painful to upgrade GLPK when a new version comes out. Generally, we try to keep patches to vendored code to a minimum for this reason.

**Comment by Antonov548** [comment]
@krlmlr I was wondering why on GHA it use vendored version of `glpk` even if it's installed.

I found that the problem is in using `R` command in configure. During the check using `rcmdcheck` R command is not available. They have it here: https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Configure-example
`Do not invoke R by plain R, Rscript or...`

I have changed it to use `"${R_HOME}/bin/R"` as it was suggested in the same webpage.

File: DESCRIPTION [hunk 4]
```diff
@@ -57,9 +57,9 @@ Suggests:
     testthat,
     vdiffr,
     withr
-LinkingTo: 
+LinkingTo:
```
Comment (krlmlr) [inline DESCRIPTION:60]:
```suggestion
LinkingTo: 
```

File: DESCRIPTION [hunk 5]
```diff
@@ -57,9 +57,9 @@ Suggests:
...
     cpp11 (>= 0.4.7)
-VignetteBuilder: 
+VignetteBuilder:
```
Comment (krlmlr) [inline DESCRIPTION:62]: The spaces here are idiomatic, unfortunately.

```suggestion
VignetteBuilder: 
```

File: rconfigure.py [hunk 6]
```diff
@@ -15,19 +15,24 @@ def open_utf8(fpath, flags):
...
+    text = f"{variable}=" + ' '.join(sources) + '\n'
 
-with open_utf8(os.path.join('src', 'sources.mk'), 'w') as f:
-     f.write(text)
+    with open_utf8(os.path.join('src' + path, f'{name}.mk'), 'w') as f:
+        f.write(text)
+
+generate_sources('/', 'sources', 'SOURCES', ignore_folders)
+generate_sources('/vendor/cigraph/vendor/glpk', 'sources-glpk', 'GLPKSOURCES', [], ('.c', '.cc', '.cpp'))
+generate_sources('/vendor/cigraph/vendor/mini-gmp', 'sources-mini-gmp', 'MINIGMPSOURCES')
```
[the remark this answers is not in the document]
Comment (krlmlr) [reply rconfigure.py:38]: Still needed here, but let's remove mini-gmp from `ignore_folders` in a separate PR.

**Review by krlmlr — APPROVED** [review]

**Pull request merged**